@ariakit/test: type() marks the field dirty before any keystroke, firing a spurious change event for readonly or keydown-prevented fields

### Current behavior

`type()` from `@ariakit/test` flags the target element as "dirty" unconditionally, before the typing loop dispatches a single keystroke ([type.ts#L60-L61](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/type.ts#L60)):

```ts
await focus(element);

// Set element dirty so blur() can dispatch a change event
element.dirty = true;
```

That flag is the sole trigger for a native `change` event when focus later leaves the element. `blur()` dispatches it directly ([blur.ts#L28-L31](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/blur.ts#L28-L31)), and `focus()` dispatches it on the previously active element before moving focus ([focus.ts#L28-L31](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/focus.ts#L28-L31)).

Real browsers fire `change` on blur only when the value actually changed while the element was focused. But because the flag is set before any keystroke, it sticks even when the interaction can never change the value. The typing loop itself enumerates these cases: a `readOnly` field or a `preventDefault()`-ed `keydown` clears `defaultAllowed` and skips the entire value-updating branch ([type.ts#L111-L138](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/type.ts#L111-L138)), and a non-text-field target (e.g. typeahead like `type("d")` on a menu item) never enters that branch at all. A fourth case is the no-op edit: Backspace/Delete in an empty input — and `press.Backspace`/`press.Delete` route straight into `type()` for text fields ([press.ts#L301-L312](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/press.ts#L301-L312)), so this happens in ordinary `press`-based tests too.

This is an internal inconsistency, not a design choice: the same package explicitly implements browser-accurate `change` semantics for click activation, firing it only when the value changed ("Real browsers fire `input`/`change` only when the value changed", [dispatch.ts#L363-L366](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/dispatch.ts#L363-L366)).

There is also a latent secondary defect in *where* the flag lands. The loop re-fetches the active element after each `keydown` because a handler may re-target focus ([type.ts#L73-L75](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/type.ts#L73)), but `dirty` was already stamped on the originally passed element — so the element that actually received the text is never marked, while the abandoned one keeps an orphaned flag.

Any test using `@ariakit/test` can therefore observe a `change` event that no browser would produce — after typing into a readonly field, after a component `preventDefault()`-ed every keystroke, after typeahead on a listbox/menu item, or after a no-op Backspace. A count-based assertion ("change fired exactly once") can then fail spuriously, and a native `change` listener (including Solid, where `onChange` maps to the native event) fires when it shouldn't. The same user scenario was verified in real Chromium with Playwright: typing into a readonly field or a keydown-prevented field and then blurring fires **0** `change` events.

### Steps to reproduce the bug

For a testing library, running a test is the consumer reproduction. Open the React template:

1. Go to https://stackblitz.com/github/ariakit/ariakit/tree/main/templates/react?file=src%2Fapp.tsx and replace `src/app.tsx` with the code below. It renders a readonly input and an input whose `onKeyDown` calls `preventDefault()`, each with a native `change` listener that increments a counter:

```tsx
import * as Ariakit from "@ariakit/react";
import type { KeyboardEvent as ReactKeyboardEvent } from "react";
import { useEffect, useRef, useState } from "react";

export default function App() {
  const readOnlyRef = useRef<HTMLInputElement>(null);
  const preventedRef = useRef<HTMLInputElement>(null);
  const [readOnlyChanges, setReadOnlyChanges] = useState(0);
  const [preventedChanges, setPreventedChanges] = useState(0);

  useEffect(() => {
    const readOnlyInput = readOnlyRef.current;
    const preventedInput = preventedRef.current;
    if (!readOnlyInput) return;
    if (!preventedInput) return;
    function onReadOnlyChange() {
      setReadOnlyChanges((count) => count + 1);
    }
    function onPreventedChange() {
      setPreventedChanges((count) => count + 1);
    }
    readOnlyInput.addEventListener("change", onReadOnlyChange);
    preventedInput.addEventListener("change", onPreventedChange);
    return () => {
      readOnlyInput.removeEventListener("change", onReadOnlyChange);
      preventedInput.removeEventListener("change", onPreventedChange);
    };
  }, []);

  function preventKeyDown(event: ReactKeyboardEvent<HTMLInputElement>) {
    event.preventDefault();
  }

  return (
    <div>
      <label htmlFor="readonly-field">Readonly field</label>
      <input ref={readOnlyRef} id="readonly-field" defaultValue="locked" readOnly />
      <label htmlFor="prevented-field">Prevented field</label>
      <input ref={preventedRef} id="prevented-field" onKeyDown={preventKeyDown} />
      <button type="button">Done</button>
      <p>
        <output role="status" aria-label="Readonly changes">{readOnlyChanges}</output>{" "}
        <output role="status" aria-label="Prevented changes">{preventedChanges}</output>
      </p>
    </div>
  );
}
```

2. Add a test (in the Ariakit repo, `app/src/sandbox/test-type-dirty-flag/test.ts`):

```ts
import { click, q, type } from "@ariakit/test";
import { expect, test } from "vitest";

function getCount(name: string) {
  return q.status.ensure(name).textContent;
}

test("typing into a readonly field fires no change event on blur", async () => {
  const field = q.textbox.ensure("Readonly field") as HTMLInputElement;
  await type("abc", field);
  await click(q.button.ensure("Done"));
  expect(field.value).toBe("locked");
  expect(getCount("Readonly changes")).toBe("0");
});

test("typing into a keydown-prevented field fires no change event on blur", async () => {
  const field = q.textbox.ensure("Prevented field") as HTMLInputElement;
  await type("abc", field);
  await click(q.button.ensure("Done"));
  expect(field.value).toBe("");
  expect(getCount("Prevented changes")).toBe("0");
});
```

3. Run `pnpm test-react test-type-dirty-flag` from the repo root. Both tests fail:

```
AssertionError: expected '1' to be '0'   // Readonly changes
AssertionError: expected '1' to be '0'   // Prevented changes
```

The exact same user scenario was verified in a real browser with Playwright (`test-browser.ts`, real clicks and key presses on the rendered page): both tests pass in Chromium, which fires `0` `change` events on blur because neither field's value ever changed.

### Expected behavior

`type()` should mark the field dirty only when a keystroke actually changes the value, so `blur()`/`focus()` dispatch `change` the way real browsers do — only when the value changed since focus. A readonly field, a field whose every `keydown` was prevented, a non-text-field typeahead target, and a no-op Backspace/Delete in an empty field should all fire **0** `change` events on blur, matching Chromium.

### Workaround

Consumer-side, in test code only. `dirty` is a plain property `@ariakit/test` stamps on the element ([`DirtiableElement = Element & { dirty?: boolean }`](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/__utils.ts#L3)), so clear it after `type()` and before the action that moves focus:

```ts
function clearDirty(element: Element & { dirty?: boolean }) {
  element.dirty = false;
}

await type("abc", field);
clearDirty(field);
await click(q.button.ensure("Done")); // no spurious `change` now
```

Verified: with this in place both tests above pass on the current `@ariakit/test`. Use it only when the interaction genuinely didn't change the value (readonly, fully prevented keydown, typeahead, no-op edit); for real edits, leave the flag alone so the legitimate `change` still fires.

### Possible solutions

Move the `dirty` stamp from "before any keystroke" to "when a keystroke actually changes the value", in `type.ts` only — `blur()`/`focus()` logic stays as is. Two subtleties: the harness dispatches `input` even for a no-op Backspace/Delete on an empty field, so a `value !== input.value` comparison is required (gating on the dispatch alone is not enough); and that comparison must be captured *before* `dispatch.compositionUpdate` runs, because its `target: { value }` is applied to the node at event-creation time, so a later comparison would always be equal during composition typing. Setting the flag inside the `isTextField` branch also writes it on the loop-local re-fetched `element`, fixing the orphaned-flag bug when a `keydown` handler re-targets focus mid-typing (the narrowed type still has `dirty` via `DirtiableElement`).

```diff
--- a/packages/ariakit-test/src/type.ts
+++ b/packages/ariakit-test/src/type.ts
@@ -57,9 +57,6 @@ export function type(

     await focus(element);

-    // Set element dirty so blur() can dispatch a change event
-    element.dirty = true;
-
     const restoreEmailInput = workAroundEmailInput(element);

     for (const char of text) {
@@ -108,6 +105,12 @@ export function type(
           value = `${firstPart}${char}${lastPart}`;
         }

+        // Whether this keystroke changes the value must be captured before any
+        // dispatch below: compositionUpdate carries `target: { value }`, which
+        // testing-library applies to the element when the event is created, so
+        // a later comparison would always be equal.
+        const valueChanged = value !== input.value;
+
         if (defaultAllowed && !input.readOnly) {
           if (inputType === "insertText") {
             defaultAllowed = await dispatch.keyPress(input, {
@@ -124,6 +127,14 @@ export function type(
             });
           }
           if (defaultAllowed) {
+            // Mark the field dirty only when the keystroke actually changes the
+            // value, so blur()/focus() dispatch `change` the way browsers do: a
+            // readonly field, a prevented keydown, a non-text-field target, or a
+            // no-op edit (e.g. Backspace in an empty input) never fires `change`.
+            if (valueChanged) {
+              element.dirty = true;
+            }
             await dispatch.input(input, {
               data: char,
               target: {
```

The `valueChanged` guard wraps only the `dirty` assignment — the `input`-event stream is untouched, keeping the blast radius minimal for consumers that assert on `input` events. The JSDoc on `blur()`/`focus()` ([blur.ts#L6-L17](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/blur.ts#L6-L17), [focus.ts#L6-L16](https://github.com/ariakit/ariakit/blob/11bd4458de6447b996dd882ea0c85eb801a6cd03/packages/ariakit-test/src/focus.ts#L6-L16)) should be tightened to say `change` fires only when typing *changed* the value. This diff is derived from the source above (not test-suite-verified here). Worth adding `@ariakit/test` unit coverage for the four suppressed cases plus one positive case asserting `change` still fires on blur after a real edit.

Since `@ariakit/test` is published and this changes an observable event, add a patch changeset (the repo's `200-` prefix for `@ariakit/test` patches):

```md
---
"@ariakit/test": patch
---

Fixed `blur` and `focus` so they no longer dispatch a `change` event after `type` interactions that never changed the field's value — readonly fields, fully prevented keydowns, typeahead on non-text-field targets, and no-op Backspace/Delete edits — matching real browser behavior. Test suites that relied on the spurious `change` event may need updating.
```

*Found by an automated multi-agent audit of the repo at commit [`11bd4458d`](https://github.com/ariakit/ariakit/tree/11bd4458de6447b996dd882ea0c85eb801a6cd03) and verified with a real-browser Playwright reproduction (user-level interactions only).*

--- a/.changeset/300-type-dirty-change.md
+++ b/.changeset/300-type-dirty-change.md
@@ -1,0 +1,5 @@
+---
+"@ariakit/test": patch
+---
+
+Fixed `type` so readonly fields, prevented keydowns, non-text-field targets, and no-op deletion keys no longer cause `blur` or `focus` to dispatch a `change` event when the value did not change.

--- a/packages/ariakit-test/readme.md
+++ b/packages/ariakit-test/readme.md
@@ -56,7 +56,7 @@ The `@ariakit/test/react` entry point renders React components for testing, and 
 function blur(element?: DirtiableElement | null): Promise<void>;
 ```
 
-Removes focus from an element, simulating a real user moving focus away from it. When no element is passed, the currently focused element (`document.activeElement`) is used. If the element was typed into since it gained focus, a `change` event is dispatched before it's blurred.
+Removes focus from an element, simulating a real user moving focus away from it. When no element is passed, the currently focused element (`document.activeElement`) is used. If typing changed the element's value since it gained focus, a `change` event is dispatched before it's blurred.
 
 Example:
 
@@ -135,7 +135,7 @@ await dispatch(q.textbox(), new Event("selectstart", { bubbles: true }));
 function focus(element: Element | null): Promise<void>;
 ```
 
-Moves focus to an element, simulating a real user focusing it. Elements that aren't focusable are ignored, and focusing the already focused element is a no-op. If another element was typed into since it gained focus, its pending `change` event is dispatched before focus moves.
+Moves focus to an element, simulating a real user focusing it. Elements that aren't focusable are ignored, and focusing the already focused element is a no-op. If typing changed another element's value since it gained focus, its pending `change` event is dispatched before focus moves.
 
 Example:
 

--- a/packages/ariakit-test/src/blur.ts
+++ b/packages/ariakit-test/src/blur.ts
@@ -6,8 +6,8 @@ import { dispatch } from "./dispatch.ts";
 /**
  * Removes focus from an element, simulating a real user moving focus away from
  * it. When no element is passed, the currently focused element
- * (`document.activeElement`) is used. If the element was typed into since it
- * gained focus, a `change` event is dispatched before it's blurred.
+ * (`document.activeElement`) is used. If typing changed the element's value
+ * since it gained focus, a `change` event is dispatched before it's blurred.
  * @example
  * ```ts
  * await type("hello", q.textbox());

--- a/packages/ariakit-test/src/focus.ts
+++ b/packages/ariakit-test/src/focus.ts
@@ -6,8 +6,8 @@ import { dispatch } from "./dispatch.ts";
 /**
  * Moves focus to an element, simulating a real user focusing it. Elements that
  * aren't focusable are ignored, and focusing the already focused element is a
- * no-op. If another element was typed into since it gained focus, its pending
- * `change` event is dispatched before focus moves.
+ * no-op. If typing changed another element's value since it gained focus, its
+ * pending `change` event is dispatched before focus moves.
  * @example
  * ```ts
  * await focus(q.textbox());

--- a/packages/ariakit-test/src/type.test.ts
+++ b/packages/ariakit-test/src/type.test.ts
@@ -1,0 +1,107 @@
+import { afterEach, expect, test, vi } from "vitest";
+import { blur } from "./blur.ts";
+import { press } from "./press.ts";
+import { type } from "./type.ts";
+
+afterEach(() => {
+  document.body.innerHTML = "";
+});
+
+function createInput(value = "") {
+  const input = document.createElement("input");
+  input.value = value;
+  document.body.append(input);
+  return input;
+}
+
+function trackChange(element: Element) {
+  const onChange = vi.fn();
+  element.addEventListener("change", onChange);
+  return onChange;
+}
+
+test("type does not dispatch change after typing into a readonly field", async () => {
+  const input = createInput("locked");
+  const onChange = trackChange(input);
+  input.readOnly = true;
+
+  await type("abc", input);
+  await blur(input);
+
+  expect(input.value).toBe("locked");
+  expect(onChange).not.toHaveBeenCalled();
+});
+
+test("type does not dispatch change after a prevented keydown", async () => {
+  const input = createInput();
+  const onChange = trackChange(input);
+  input.addEventListener("keydown", (event) => {
+    event.preventDefault();
+  });
+
+  await type("abc", input);
+  await blur(input);
+
+  expect(input.value).toBe("");
+  expect(onChange).not.toHaveBeenCalled();
+});
+
+test("type does not dispatch change after typing on a non-text field", async () => {
+  const target = document.createElement("div");
+  const onChange = trackChange(target);
+  target.tabIndex = 0;
+  target.textContent = "Delta";
+  document.body.append(target);
+
+  await type("d", target);
+  await blur(target);
+
+  expect(onChange).not.toHaveBeenCalled();
+});
+
+test.each([
+  { name: "Backspace", pressKey: press.Backspace },
+  { name: "Delete", pressKey: press.Delete },
+])(
+  "press.$name does not dispatch change after a no-op edit",
+  async ({ pressKey }) => {
+    const input = createInput();
+    const onChange = trackChange(input);
+
+    await pressKey(input);
+    await blur(input);
+
+    expect(input.value).toBe("");
+    expect(onChange).not.toHaveBeenCalled();
+  },
+);
+
+test("type dispatches change on blur after editing a text field", async () => {
+  const input = createInput("a");
+  const onChange = trackChange(input);
+  input.setSelectionRange(input.value.length, input.value.length);
+
+  await type("b", input);
+  await blur(input);
+
+  expect(input.value).toBe("ab");
+  expect(onChange).toHaveBeenCalledOnce();
+});
+
+test("type marks the text field that receives text after focus moves", async () => {
+  const firstInput = createInput();
+  const secondInput = createInput();
+  const onFirstInputChange = trackChange(firstInput);
+  const onSecondInputChange = trackChange(secondInput);
+  firstInput.addEventListener("keydown", () => {
+    secondInput.focus();
+  });
+
+  await type("a", firstInput);
+  await blur(secondInput);
+
+  expect(firstInput.value).toBe("");
+  expect(secondInput.value).toBe("a");
+  expect(onFirstInputChange).not.toHaveBeenCalled();
+  expect(onSecondInputChange).toHaveBeenCalledOnce();
+});

--- a/packages/ariakit-test/src/type.ts
+++ b/packages/ariakit-test/src/type.ts
@@ -62,9 +62,6 @@ export function type(
 
     await focus(element);
 
-    // Set element dirty so blur() can dispatch a change event
-    element.dirty = true;
-
     const restoreEmailInput = workAroundEmailInput(element);
 
     for (const char of text) {
@@ -80,7 +77,7 @@ export function type(
       element = getActiveElement(element) || element;
 
       if (isTextField(element)) {
-        const input = element as TextField;
+        const input = element as DirtiableElement & TextField;
         const [start, end] = [
           input.selectionStart ?? 0,
           input.selectionEnd ?? 0,
@@ -113,6 +110,10 @@ export function type(
           value = `${firstPart}${char}${lastPart}`;
         }
 
+        // Capture this before dispatch.compositionUpdate(), which applies
+        // target.value when creating the event.
+        const valueChanged = value !== input.value;
+
         if (defaultAllowed && !input.readOnly) {
           if (inputType === "insertText") {
             defaultAllowed = await dispatch.keyPress(input, {
@@ -129,6 +130,9 @@ export function type(
             });
           }
           if (defaultAllowed) {
+            if (valueChanged) {
+              input.dirty = true;
+            }
             await dispatch.input(input, {
               data: char,
               target: {

